### PR TITLE
300/provider foreach

### PR DESCRIPTION
**Pull request by Andrew-Hayes**
Resolves # [300](https://github.com/opentofu/opentofu/issues/300)

This PR adds for_each support to providers. To achieve that we have made a few changes

- We evaluate a provider and if it has a for_each statement we create multiple instances of the provider based on the for_each iteration.
- Within the provider we now have a ProviderBlock and Provider. The ProviderBlock is populated from the provider block in the hcl and then we evaluate that. This is then translated to 1 or more "Provider" 's that contain the instance specific data. 
- We have moved the evaluation checks for the for_each and count of a resource into a "lang" sub package. So it can be re-used. We then make use of this for provider for_each evaluations. 
- Added unit tests for the new changes

## Target Release

1.9.0

## Checklist

- [x] I have read the [contribution guide](https://github.com/opentofu/opentofu/blob/main/CONTRIBUTING.md).
- [x] I have not used an AI coding assistant to create this PR.
- [x] I have written all code in this PR myself OR I have marked all code I have not written myself (including modified code, e.g. copied from other places and then modified) with a comment indicating where it came from.

### Go checklist

 

- [x] I have run golangci-lint on my change and receive no errors relevant to my code.
- [x] I have run existing tests to ensure my code doesn't break anything.
- [x] I have added tests for all relevant use cases of my code, and those tests are passing.
- [x] I have only exported functions, variables and structs that should be used from other packages.
- [x] I have added meaningful comments to all exported functions, variables, and structs.

### Website/documentation checklist

- [ ] I have locally started the website as [described here](https://github.com/opentofu/opentofu/blob/main/website/README.md) and checked my changes.

### CODE REVIEW

File: internal/configs/resource.go [hunk 1]
```diff
@@ -668,6 +668,24 @@ func decodeProviderConfigRef(expr hcl.Expression, argName string) (*ProviderConf
 	expr, shimDiags = shimTraversalInString(expr, false)
 	diags = append(diags, shimDiags...)
 
+	if idxexpr, ok := expr.(*hclsyntax.IndexExpr); ok {
```
Comment (cam72cam) [inline internal/configs/resource.go:671]: I'm not sure why these changes are in this PR

File: internal/configs/provider.go [hunk 2]
```diff
@@ -70,17 +77,25 @@ func decodeProviderBlock(block *hcl.Block) (*Provider, hcl.Diagnostics) {
...
-		}
+	if attr, exists := content.Attributes["for_each"]; exists {
+		provider.ForEach = attr.Expr
+	}
+
+	if attr, exists := content.Attributes["count"]; exists {
+		provider.Count = attr.Expr
+	}
+
+	if provider.Count != nil && provider.ForEach != nil && provider.AliasExpr != nil {
```
Comment (cam72cam) [inline internal/configs/provider.go:92]: This expression is incorrect

File: internal/configs/provider.go [hunk 3]
```diff
@@ -145,6 +160,109 @@ func decodeProviderBlock(block *hcl.Block) (*Provider, hcl.Diagnostics) {
...
+		diags = append(diags, evalDiags.ToHCL()...)
+		if evalDiags.HasErrors() {
+			return nil, diags
+		}
+
+		var out []*Provider
+		for k, v := range forVal {
+			v := v
+			iter := *p
+			iter.Alias = k
```
Comment (cam72cam) [inline internal/configs/provider.go:195]: We probably want to check `hclsyntax.ValidIdentifier(alias)`, not sure how the numeric count will handle that name.0 for example.

File: internal/configs/provider.go (hunk 2 above)
Comment (Andrew-Hayes) [reply internal/configs/provider.go:92]: yeah, i thought that when writing it. now to remember the logic lecture from first year of uni.

Comment (cam72cam) [reply internal/configs/provider.go]: I'd recommend something like:
```
if (alias != nil && count != nil) || (alias != nil && for_each != nil) || (count != nill && for_each != nil) {
  // common error message
}
```

File: internal/tofu/eval_iteration_test.go
Comment (cam72cam) [inline internal/tofu/eval_iteration_test.go:1]: I wonder if it makes sense to migrate some/all these tests to the eval_checks package?

File: internal/configs/provider_iteration_test.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+	if name, exists := mod.ProviderLocalNames[p]; !exists {
+		t.Fatal("provider FQN hashicorp/local not found")
+	} else {
+		if name != "local" {
+			t.Fatalf("provider localname mismatch: got %s, want local", name)
+		}
+	}
+
+	if len(mod.ProviderConfigs) != 3 {
+		t.Fatal("incorrect numver of providers")
```
Comment (cam72cam) [inline internal/configs/provider_iteration_test.go:30]: number

File: internal/tofu/eval_iteration_test.go
Comment (Andrew-Hayes) [reply internal/tofu/eval_iteration_test.go:1]: yeah it does. some are tightly coupled though, even more so than the main code. but is def worth the effort.

File: internal/configs/provider.go [hunk 5]
```diff
@@ -290,3 +333,89 @@ func checkProviderNameNormalized(name string, declrange hcl.Range) hcl.Diagnosti
...
 	return diags
 }
+
+func (p *Provider) generateForEachProviders(eval *StaticEvaluator) ([]*Provider, hcl.Diagnostics) {
+	var diags hcl.Diagnostics
+	if eval == nil {
+		return nil, diags.Append(&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Iteration not allowed in test files",
+			Detail:   "for_each was declared as an provider attribute in a test file",
```
Comment (ollevche) [inline internal/configs/provider.go:343]:
```suggestion
			Detail:   "for_each was declared as a provider attribute in a test file",
```

File: internal/configs/provider.go [hunk 6]
```diff
@@ -290,3 +333,89 @@ func checkProviderNameNormalized(name string, declrange hcl.Range) hcl.Diagnosti
...
+			Subject:  p.AliasExpr.Range().Ptr(),
+		})
+	}
+
+	foeachRefsFunc := func(refs []*addrs.Reference) (*hcl.EvalContext, tfdiags.Diagnostics) {
```
Comment (ollevche) [inline internal/configs/provider.go:348]: foeachRefsFunc -> forEachRefsFunc

File: internal/configs/provider.go [hunk 7]
```diff
@@ -290,3 +333,89 @@ func checkProviderNameNormalized(name string, declrange hcl.Range) hcl.Diagnosti
...
 }
+
+func (p *Provider) generateForEachProviders(eval *StaticEvaluator) ([]*Provider, hcl.Diagnostics) {
+	var diags hcl.Diagnostics
+	if eval == nil {
+		return nil, diags.Append(&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Iteration not allowed in test files",
+			Detail:   "for_each was declared as an provider attribute in a test file",
+			Subject:  p.AliasExpr.Range().Ptr(),
```
Comment (ollevche) [inline internal/configs/provider.go:344]:
```suggestion
			Subject:  p.ForEach.Range().Ptr(),
```

File: internal/configs/provider.go [hunk 8]
```diff
@@ -290,3 +333,89 @@ func checkProviderNameNormalized(name string, declrange hcl.Range) hcl.Diagnosti
...
+		return nil, diags
+	}
+
+	var out []*Provider
+	for k, v := range forVal {
+		if !hclsyntax.ValidIdentifier(k) {
+			return nil, diags.Append(&hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "Invalid Identifier",
+				Detail:   fmt.Sprintf("The provided identifier %s is invalid", k),
```
Comment (ollevche) [inline internal/configs/provider.go:370]: nit: Invalid for_each key identifier?

File: internal/configs/provider.go [hunk 9]
```diff
@@ -290,3 +333,89 @@ func checkProviderNameNormalized(name string, declrange hcl.Range) hcl.Diagnosti
...
+	return out, diags
+}
+
+func (p *Provider) generateCountProviders(eval *StaticEvaluator) ([]*Provider, hcl.Diagnostics) {
+	var diags hcl.Diagnostics
+	if eval == nil {
+		return nil, diags.Append(&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Iteration not allowed in test files",
+			Detail:   "for_each was declared as an provider attribute in a test file",
```
Comment (ollevche) [inline internal/configs/provider.go:391]:
```suggestion
			Detail:   "count was declared as a provider attribute in a test file",
```

File: internal/configs/provider.go [hunk 10]
```diff
@@ -290,3 +333,89 @@ func checkProviderNameNormalized(name string, declrange hcl.Range) hcl.Diagnosti
...
+			Subject:  p.AliasExpr.Range().Ptr(),
```
Comment (ollevche) [inline internal/configs/provider.go:392]:
```suggestion
			Subject:  p.Count.Range().Ptr(),
```

File: internal/configs/provider.go [hunk 11]
```diff
@@ -290,3 +333,89 @@ func checkProviderNameNormalized(name string, declrange hcl.Range) hcl.Diagnosti
...
+	}
+
+	foeachRefsFunc := func(refs []*addrs.Reference) (*hcl.EvalContext, tfdiags.Diagnostics) {
+		var diags tfdiags.Diagnostics
+		evalContext, evalDiags := eval.EvalContext(StaticIdentifier{
+			Module:    eval.call.addr,
+			Subject:   fmt.Sprintf("provider.%s.for_each", p.Name),
+			DeclRange: p.ForEach.Range(),
+		}, refs)
+		return evalContext, diags.Append(evalDiags)
```
Comment (ollevche) [inline internal/configs/provider.go:355]: Seems like we will have duplicated diags if we append them here and in `EvaluateForEachExpression`. The same for count

File: internal/configs/provider.go [hunk 12]
```diff
@@ -290,3 +333,89 @@ func checkProviderNameNormalized(name string, declrange hcl.Range) hcl.Diagnosti
...
+	var out []*Provider
+	countVal, evalDiags := evalchecks.EvaluateCountExpression(p.Count, countEvalFunc)
+	diags = append(diags, evalDiags.ToHCL()...)
+	if evalDiags.HasErrors() {
+		return nil, diags
+	}
+
+	for i := 0; i < countVal; i++ {
+		iter := *p
+		iter.Alias = fmt.Sprintf("[%d]", i)
```
Comment (ollevche) [inline internal/configs/provider.go:414]: Do we want count providers to be available only via square brackets? Should we let the caller decide?

File: internal/eval_checks/diagnostics.go [hunk 13]
```diff
@@ -3,7 +3,7 @@
 // Copyright (c) 2023 HashiCorp, Inc.
 // SPDX-License-Identifier: MPL-2.0
 
-package tofu
+package evalchecks
```
Comment (ollevche) [inline internal/eval_checks/diagnostics.go:6]: can we make the package name the same as the folder name?

**Review by ollevche — COMMENTED** [review]
Overall it looks pretty solid, left a couple of questions and suggestions. Thank you!

File: internal/configs/provider.go [hunk 14]
```diff
@@ -23,7 +25,12 @@ type Provider struct {
 	Name       string
 	NameRange  hcl.Range
 	Alias      string
-	AliasRange *hcl.Range // nil if no alias set
+	AliasExpr  hcl.Expression // nil if no alias set
+	AliasRange *hcl.Range     // nil if no alias set
+	ForEach    hcl.Expression
+	EachValue  *cty.Value
+	Count      hcl.Expression
+	CountIndex *cty.Value
```
Comment (Evi1Pumpkin) [inline internal/configs/provider.go:33]: Two things that quite bother me with this implementation:
1. This struct is no longer as described in the comment above: `Provider represents a "provider" block in a module or file.` It sometimes represents a block and sometimes a provider "instance".
2. Now, it holds two responsibilities and changes based on the application's lifecycle. If `tofu` is in the stage of reading the configuration, this struct represents the `provider` block. If `tofu` is building the module tree and populating the providers (after calling `decodeStaticFields()`), the struct now describes a specific "instance" of the provider.

This implementation is quite confusing and different from the practices in the tofu repository. Maybe we should consider adding a different struct that will include the `EachValue` and `CountIndex`?

**Review by Evi1Pumpkin — COMMENTED** [review]
Overall looking excellent, left a few comments

File: internal/configs/provider.go (hunk 14 above)
Comment (cam72cam) [reply internal/configs/provider.go:33]: 1. It's more so a "provider block" vs "provider block with instance data"
2. This is true, the expansion here feels a bit weird IMO and I noticed that during the prototype phase.  This change is the "low churn" option, but adds cognitive load to understanding provider configs in the configs package. 

There are technically 3 levels of config parsing that are all mixed together:
1. hcl/json text -> configs.File (HCL loading)
2. configs.File -> configs.Module (merging files)
3. configs.Module -> configs.Module (static evaluation)

Even before static evaluation was added, there were all sorts of odd validation bugs all throughout the code when dealing with overrides.  Static evaluation shines more of a light on this.

In this specific scenario, we would need:
```go

// File parsing produces
type configs.ProviderBlock struct {
  // 1-1 mapping to HCL
}

// Module struct from Files (normal+override) produces
type configs.ProviderData struct {
  // Closer to existing struct
  ForEach hcl.Expression
  Count hcl.Expression
}
// Validation is performed when constructing ProviderData

// func (ProviderData) expand(StaticContext) []Provider

// Expanded into one or more
type configs.Provider struct {
  // Matches existing struct
  
  ForEachValue *cty.Value
  CountValue *cty.Value
  // OR
  InstanceData instances.RepetitionData
}
```


For the purposes of this PR, we could merge the concept of ProviderBlock and ProviderData to reduce the potential set of changes as it is a less important distinction IMO.

File: internal/eval_checks/diagnostics.go (hunk 13 above)
Comment (Andrew-Hayes) [reply internal/eval_checks/diagnostics.go:6]: oh yeah. i just lazily threw the package here until we figured out where it should go.. then never did..

File: internal/configs/provider.go (hunk 14 above)
Comment (ronnyorot) [reply internal/configs/provider.go:33]: I agree ^
Also, I prefer the usage of `InstanceData instances.RepetitionData` over `ForEachValue` and `CountValue`.

Comment (Evi1Pumpkin) [reply internal/configs/provider.go:33]: I agree ^
Also, I prefer the usage of InstanceData instances.RepetitionData over ForEachValue and CountValue.

Comment (ollevche) [reply internal/configs/provider.go:33]: We already track some of the modernization/refactoring issues and I would suggest we document that down to address later. Both @Evi1Pumpkin and @cam72cam arguments sound pretty reasonable.

Breaking down config package structures into different stages such as loading (raw) and evaluation (ready for graph processing) makes perfect sense to me.

File: internal/configs/provider.go [hunk 15]
```diff
@@ -70,17 +77,25 @@ func decodeProviderBlock(block *hcl.Block) (*Provider, hcl.Diagnostics) {
...
+	if attr, exists := content.Attributes["count"]; exists {
+		provider.Count = attr.Expr
+	}
+
+	if (provider.AliasExpr != nil && provider.Count != nil) || (provider.AliasExpr != nil && provider.ForEach != nil) || (provider.Count != nil && provider.ForEach != nil) {
+		diags = append(diags, &hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  `Invalid combination of "alias", "count" and "for_each"`,
+			Detail:   `The "alias", "count" and "for_each" arguments are mutually-exclusive, only one may be used.`,
+			Subject:  provider.AliasExpr.Range().Ptr(),
```
[the remark this answers is not in the document]
Comment (Andrew-Hayes) [reply internal/configs/provider.go:97]: funny i did have them split but Christian advocated for a combined error message

File: internal/configs/provider.go (hunk 12 above)
Comment (Andrew-Hayes) [reply internal/configs/provider.go:414]: interesting. so i had made changes to resource.go so i could reference a provider like so: 

```
provider = local.0
```

this made use of https://github.com/opentofu/opentofu/blob/main/internal/addrs/instance_key.go#L66-L68 

this meant that it had to be wrapped in square brackets to match. However thats all been remove as it steps on the toes of what you are working on. So not sure if i still need them 🤔

File: internal/configs/provider.go [hunk 16]
```diff
@@ -154,13 +204,6 @@ func (p *Provider) Addr() addrs.LocalProviderConfig {
 	}
 }
 
-func (p *Provider) moduleUniqueKey() string {
```
[the remark this answers is not in the document]
Comment (Andrew-Hayes) [reply internal/configs/provider.go:157]: i had initially copied it from the original branch without thinking about it. So it is perhaps not needed, but as Addr().StringCompact() already exists is it duplicate code to have both? 🤔

[the remark this answers is not in the document]
Comment (Evi1Pumpkin) [reply internal/configs/provider.go:157]: I don't mind, we can leave it like this. I was trying to understand the reason behind that and if I misunderstood or missed something. Thanks for the clarification!

File: internal/configs/provider.go (hunk 14 above)
Comment (Andrew-Hayes) [reply internal/configs/provider.go:33]: We discussed this offline. The agreement was to go with @cam72cam 's suggestion and do 2 structs: one that captures the understanding of "pre-evaluation" and one that is "post evaluation". Then we will revisit in future with an RFC on larger refactoring work.

File: internal/configs/provider.go (hunk 11 above)
Comment (Andrew-Hayes) [reply internal/configs/provider.go:355]: this func is called in EvaluateForEachExpression. So EvaluateForEachExpression returns this diags as it's own.

File: internal/configs/provider.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (ollevche) [reply internal/configs/provider.go:97]: It is nit, you can just ignore that part of the comment

File: internal/configs/provider.go (hunk 11 above)
Comment (ollevche) [reply internal/configs/provider.go:355]: My bad, I thought `diags` variable is from the outer scope. Thanks!

File: internal/configs/provider.go [hunk 17]
```diff
@@ -290,3 +354,104 @@ func checkProviderNameNormalized(name string, declrange hcl.Range) hcl.Diagnosti
 	}
 	return diags
 }
+
+func (p *ProviderBlock) toProvider() Provider {
+	return Provider{
+		ProviderCommon: ProviderCommon{
```
Comment (cam72cam) [inline internal/configs/provider.go:360]: return Provider{ProviderCommon: p.ProviderCommon}

File: internal/configs/provider.go [hunk 18]
```diff
@@ -154,11 +219,12 @@ func (p *Provider) Addr() addrs.LocalProviderConfig {
 	}
 }
 
-func (p *Provider) moduleUniqueKey() string {
-	if p.Alias != "" {
-		return fmt.Sprintf("%s.%s", p.Name, p.Alias)
+// Addr returns the address of the receiving provider configuration, relative
+// to its containing module.
+func (p *ProviderBlock) Addr() addrs.LocalProviderConfig {
```
Comment (cam72cam) [inline internal/configs/provider.go:224]: This function is no longer used

File: internal/configs/provider.go [hunk 19]
```diff
@@ -13,17 +13,14 @@ import (
...
 
-// Provider represents a "provider" block in a module or file. A provider
-// block is a provider configuration, and there can be zero or more
-// configurations for each actual provider.
-type Provider struct {
-	Name       string
-	NameRange  hcl.Range
-	Alias      string
-	AliasRange *hcl.Range // nil if no alias set
+type ProviderCommon struct {
```
Comment (ollevche) [inline internal/configs/provider.go:21]: I understand, we already discussed this previously and I do not have a strong opinion here.

But...

Do we need `ProviderCommon`? It is not useful by itself and doesn't say anything about what it really is.

We could move `ProviderCommon` fields to `ProviderBlock` and tweak a `Provider` struct to include a `ConfigBlock` field of type `ProviderBlock` (instead of embedding). This looks good to me since we have both clean separation and source configuration values.

File: internal/configs/provider.go [hunk 20]
```diff
@@ -290,3 +321,76 @@ func checkProviderNameNormalized(name string, declrange hcl.Range) hcl.Diagnosti
...
+				Detail:   fmt.Sprintf("The provided identifier %s is invalid", k),
+				Subject:  p.ForEach.Range().Ptr(),
+			})
+		}
+
+		v := v
+		iter := Provider{ProviderCommon: p.ProviderCommon}
+		iter.Alias = k
+		iter.InstanceData.EachValue = v
+		out = append(out, &iter)
```
Comment (ollevche) [inline internal/configs/provider.go:367]: nit:

```suggestion
		out = append(out, &Provider{
			ProviderCommon:		p.ProviderCommon,
			Alias:			k,
			InstanceData:		instances.RepetitionData{
				EachValue:	v,
			},
		})
```

File: internal/configs/test_file.go [hunk 21]
```diff
@@ -396,10 +397,22 @@ func loadTestFile(body hcl.Body) (*TestFile, hcl.Diagnostics) {
...
+				diags = append(diags, decodeDiags...)
+				if decodeDiags.HasErrors() {
+					continue
+				}
+				providers, diagsStatic := providerBlock.decodeStaticFields(nil)
+				diags = append(diags, diagsStatic...)
+				if diagsStatic.HasErrors() {
+					continue
+				}
+				for _, provider := range providers {
```
Comment (ollevche) [inline internal/configs/test_file.go:413]: Creating a provider list out of a provider block with a nil static eval should always result in a single provider, right?

nit: should we just make it `providers[0]` for clarity?

File: internal/evalchecks/eval_for_each.go [hunk 22]
```diff
@@ -2,33 +2,40 @@
 // SPDX-License-Identifier: MPL-2.0
 // Copyright (c) 2023 HashiCorp, Inc.
 // SPDX-License-Identifier: MPL-2.0
+package evalchecks
 
-package tofu
+// TODO this is only here until i figure out where it should be
```
Comment (ollevche) [inline internal/evalchecks/eval_for_each.go:7]: Is it still relevant?

**Review by ollevche — COMMENTED** [review]
Looks great, thank you! Left a couple of questions and several nit suggestions. Feel free to ignore them if you think they are irrelevant.

File: internal/configs/provider.go (hunk 19 above)
Comment (Andrew-Hayes) [reply internal/configs/provider.go:21]: i would tend to agree, i'm not a fan of embedding. But i'm ok with what we have

File: internal/evalchecks/eval_for_each.go (hunk 22 above)
Comment (Andrew-Hayes) [reply internal/evalchecks/eval_for_each.go:7]: kind of. I still feel like there is maybe somewhere better it should be? but no-one seems to mind it being here...

File: internal/configs/config.go [hunk 23]
```diff
@@ -966,15 +965,16 @@ func (c *Config) transformProviderConfigsForTest(run *TestRun, file *TestFile) (
...
+				ProviderCommon: ProviderCommon{
+					Name:          ref.InChild.Name,
+					NameRange:     ref.InChild.NameRange,
+					Version:       testProvider.Version,
+					Config:        testProvider.Config,
+					DeclRange:     testProvider.DeclRange,
+					IsMocked:      testProvider.IsMocked,
+					MockResources: testProvider.MockResources,
+				},
+				Alias: ref.InChild.Alias,
```
[the remark this answers is not in the document]
Comment (Andrew-Hayes) [reply internal/configs/config.go:977]: The range is part of the block as it's linked to where the expression lives. It didn't look like it was used for anything once we have the instance data etc. but I could be wrong.. 👀

[the remark this answers is not in the document]
Comment (ollevche) [reply internal/configs/config.go:977]: Yep, seems like it is not used anywhere. We can add it back anytime if needed though

File: internal/configs/provider.go [hunk 24]
```diff
@@ -290,3 +325,79 @@ func checkProviderNameNormalized(name string, declrange hcl.Range) hcl.Diagnosti
 	}
 	return diags
 }
+
+func (p *ProviderBlock) generateForEachProviders(eval *StaticEvaluator) ([]*Provider, hcl.Diagnostics) {
```
Comment (Evi1Pumpkin) [inline internal/configs/provider.go:329]: Can we keep the methods of the `ProviderBlock` struct close to each other and to the struct itself?
Or else it's a bit confusing when there are similar structs in the same file with mixed functions between them (also relevant for `providerBlockSchema`, which is only relevant to the `ProviderBlock`)

File: internal/evalchecks/eval_for_each.go (hunk 22 above)
Comment (cam72cam) [reply internal/evalchecks/eval_for_each.go:7]: I suggested maybe in the lang package.  If it does not fit there, I think leaving it as it's own package works well enough for now.

File: internal/configs/test_file.go [hunk 25]
```diff
@@ -396,10 +397,22 @@ func loadTestFile(body hcl.Body) (*TestFile, hcl.Diagnostics) {
...
+				_, decodeDiags := providerBlock.decodeStaticFields(nil)
+				diags = append(diags, decodeDiags...)
+				if decodeDiags.HasErrors() {
+					continue
+				}
+				providers, diagsStatic := providerBlock.decodeStaticFields(nil)
+				diags = append(diags, diagsStatic...)
+				if diagsStatic.HasErrors() {
+					continue
+				}
```
[the remark this answers is not in the document]
Comment (Andrew-Hayes) [reply internal/configs/test_file.go:412]: no idea how I missed this.. haha

File: internal/evalchecks/eval_for_each.go (hunk 22 above)
Comment (Andrew-Hayes) [reply internal/evalchecks/eval_for_each.go]: import cycled in the lang package itself. but I moved it into a sub package and it felt like it might be more at home there?

File: internal/configs/provider.go (hunk 24 above)
Comment (Andrew-Hayes) [reply internal/configs/provider.go:334]: ok, new layout has the following reasons: 

- Schema at the bottom due to convention. 
- ProviderBlock functions above that to keep them grouped
- Provider functions above that to keep struct methods together
- Private functions above that
- Public functions above that. because that felt right, but is probably some java convention that I remember from the dark days.

File: internal/evalchecks/eval_for_each.go (hunk 22 above)
Comment (cam72cam) [reply internal/evalchecks/eval_for_each.go]: Sounds good to me!

**Review by ollevche — DISMISSED** [review]

File: internal/configs/provider.go [hunk 26]
```diff
@@ -95,17 +200,8 @@ func decodeProviderBlock(block *hcl.Block) (*Provider, hcl.Diagnostics) {
...
-		if attr, exists := content.Attributes[name]; exists {
-			diags = append(diags, &hcl.Diagnostic{
-				Severity: hcl.DiagError,
-				Summary:  "Reserved argument name in provider block",
-				Detail:   fmt.Sprintf("The provider argument name %q is reserved for use by OpenTofu in a future version.", name),
-				Subject:  &attr.NameRange,
-			})
-		}
-	}
+	reserveredDiags := checkReserverNames(content)
```
Comment (cam72cam) [inline internal/configs/provider.go:203]: checkReservedNames

File: internal/configs/provider.go [hunk 27]
```diff
@@ -154,104 +294,108 @@ func (p *Provider) Addr() addrs.LocalProviderConfig {
...
-	if len(traversal) > 2 {
-		diags = diags.Append(&hcl.Diagnostic{
-			Severity: hcl.DiagError,
-			Summary:  "Invalid provider configuration address",
-			Detail:   "Extraneous extra operators after provider configuration address.",
-			Subject:  traversal[2:].SourceRange().Ptr(),
-		})
+	forEachRefsFunc := func(refs []*addrs.Reference) (*hcl.EvalContext, tfdiags.Diagnostics) {
+		var diags tfdiags.Diagnostics
+		evalContext, evalDiags := eval.EvalContext(StaticIdentifier{
```
Comment (cam72cam) [inline internal/configs/provider.go:347]: This can just be a return, don't need additional vars here

File: internal/configs/provider.go [hunk 28]
```diff
@@ -154,104 +294,108 @@ func (p *Provider) Addr() addrs.LocalProviderConfig {
...
-// then the returned address is invalid.
-func ParseProviderConfigCompactStr(str string) (addrs.LocalProviderConfig, tfdiags.Diagnostics) {
-	var diags tfdiags.Diagnostics
+	var out []*Provider
+	for k, v := range forVal {
+		if !hclsyntax.ValidIdentifier(k) {
+			return nil, diags.Append(&hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "Invalid for_each key Identifier",
+				Detail:   fmt.Sprintf("The provided identifier %s is invalid", k),
```
Comment (cam72cam) [inline internal/configs/provider.go:367]: Let's make this a better error message, maybe include badIdentifierDetail and rename "identifier" to  alias

File: internal/configs/provider.go (hunk 27 above)
Comment (Andrew-Hayes) [reply internal/configs/provider.go:347]: it needs them because EvalContext returns `hcl.Diagnostics` but we need `tfdiags.Diagnostics`. So have to do that append magic.

File: internal/configs/test_file.go [hunk 29]
```diff
@@ -396,10 +397,17 @@ func loadTestFile(body hcl.Body) (*TestFile, hcl.Diagnostics) {
...
 			diags = append(diags, providerDiags...)
-			if provider != nil {
-				tf.Providers[provider.moduleUniqueKey()] = provider
+			if providerBlock != nil {
+				providers, diagsStatic := providerBlock.decodeStaticFields(nil)
+				diags = append(diags, diagsStatic...)
+				if diagsStatic.HasErrors() {
+					continue
+				}
+				// providers only ever has 1 element for nul static fields.
```
Comment (cam72cam) [inline internal/configs/test_file.go:408]: typos

File: internal/configs/test_file.go [hunk 30]
```diff
@@ -396,10 +397,17 @@ func loadTestFile(body hcl.Body) (*TestFile, hcl.Diagnostics) {
...
+				provider := providers[0]
```
Comment (cam72cam) [inline internal/configs/test_file.go:409]: Even though the comment is currently true, let's do a for-loop instead just to be safe against future enhancements

File: internal/configs/provider.go (hunk 27 above)
Comment (cam72cam) [reply internal/configs/provider.go:344]: Ah, my mistake!

File: internal/configs/test_file.go (hunk 21 above)
Comment (cam72cam) [reply internal/configs/test_file.go:408]: I asked to keep it as a loop to future proof it.

**Review by cam72cam — APPROVED** [review]
Approving with the caveat that additional tests and documentation will be coming in the larger #300 feature branch.
